### DISCUSSION TITLE
Link to new a different pdf in a new window

### DISCUSSION THREAD

**Discussion by chapmanwilliam** [pymupdf/PyMuPDF#3129] [Looking for help] [answered]
Dear All, 

I've been struggling to insert a link (page.insert_link(link_dict)) that opens a different file in **a new window.**. I know it's possible because I have done it using Evermap software and looked at the xref data for the link:

  A ('xref', '67 0 R')
  	D ('array', '[22/XYZ 0 792 0]')
  	F ('xref', '68 0 R')
   		F ('string', 'buildingofficeaddins.pdf')
   		Type ('name', '/Filespec')
   		NewWindow ('bool', 'true')
   		S ('name', '/GoToR')
  BS ('dict', '<</S/U/Type/Border/W 1>>')
  Border ('array', '[0 0 1]')
  H ('name', '/I')
  M ('string', 'D:20240204201504Z')
  Rect ('array', '[72 756.568 83.077 771.286]')
  Subtype ('name', '/Link')
  Type ('name', '/Annot')

However, I have not worked out how to do this using pymudf. The xref data looks like this instead:

A ('dict', '<</S/GoToR/D[52/XYZ 0 0 0]/F<</F(example.pdf)/UF(example.pdf)/Type/Filespec>>>>')
Rect ('array', '[10 37 570 43]')
BS ('dict', '<</W 0>>')
Subtype ('name', '/Link')
NM ('string', 'fitz-L64')

Any tips gratefully received.

William Chapman

**Comment by JorjMcKie** [comment]
This does it:
```python
import fitz

doc = fitz.open("input.pdf")
page = doc[0]
r = fitz.Rect(...)
link = {
    "from": r,
    "kind": fitz.LINK_GOTOR,
    "file": path,  # target path
    "page": 0,  # target page number
    "to": fitz.Point(72, 800),  # position on target page in PDF (!) coords
}
page.insert_link(link)
doc.save("output.pdf")
```

**Comment by chapmanwilliam** [comment]
Dear Jorj,

No, it does not, unfortunately. Your code creates a link to another file,
no problem. However, when I click on the link it closes down the
originating file and opens up the new one. I want it to keep the
original file open and open a *new window.*

Is that possible?

William

**Reply by JorjMcKie** [reply]
Which PDF viewer do you use? Adobe?
Yes, Adobe closes the original file - or even crashes!

The link works as desired in:
* evince (linux)
* PDF XChange viewer
* SumatraPDF
* Foxit Reader
* Nitro PDF viewer

**Reply by chapmanwilliam** [reply]
That's a pity. I am an Adobe DC user and no chance of my switching because of this. I imagine many people are the same.

All it seems to be missing is a 'NewWindow' flag. I have successfullly taken a link (created by Evermap) from an Adobe file and switched the 'NewWindow' key flag to 'false' and then back again to 'true' using your x_ref_set_key method. Unfortunately, the links generated by pymupdf do not have a NewWindow key. Why is that?

I examined your getLinkText method that references standard txt for the link annotation. I tentatively tried the following to override and amend the txt to add a NewWindow flag but it had no effect:

            txt = annot_skel["gotor1"]  # annot_gotor
            txt1= '<</A<</S/GoToR/D[%i /XYZ %g %g %g]/F<</F(%s)/UF(%s)/Type/Filespec/NewWindow/true>>>>/Rect[%s]/BS<</W 0>>/Subtype/Link>>'
            txt=txt1

**Reply by JorjMcKie** [reply]
> do not have a NewWindow key. Why is that?

... because - as per the PDF spec - we leave that to the default behavior. 
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/fb20c0be-a220-4efb-ab5f-eb3a97c59c43)

BTW your hack cannot work because you used "true" like a PDF **_name_** (prefixed it with a "/"). Correct hack works like this:
```python
print(doc.xref_object(57))
<<
  /A <<
    /S /GoToR
    /D [ 0 /XYZ 72 800 0 ]
    /F <<
      /F (1text.pdf)
      /UF (1text.pdf)
      /Type /Filespec
    >>
  >>
  /Rect [ 70.8238 739.99 117.595 755.95 ]
  /BS <<
    /W 0
  >>
  /Subtype /Link
  /NM (fitz-L0)
>>
doc.xref_set_key(57,"A/F/NewWindow","true")
print(doc.xref_object(57))
<<
  /A <<
    /S /GoToR
    /D [ 0 /XYZ 72 800 0 ]
    /F <<
      /F (1text.pdf)
      /UF (1text.pdf)
      /Type /Filespec
      /NewWindow true
    >>
  >>
  /Rect [ 70.8238 739.99 117.595 755.95 ]
  /BS <<
    /W 0
  >>
  /Subtype /Link
  /NM (fitz-L0)
>>
```
This does not change my Adobe's behavior (still quits unexpectedly - however works every other time it seems).

**Reply by chapmanwilliam** [reply]
Thank you. However, even with your hack the behaviour does not change. But it doesn't crash!

My link looks like this with your hack:

/A <<
    /S /GoToR
    /D [ 54 /XYZ 0 0 0 ]
    /F <<
      /F (C:/Users/samsu/Dropbox/My documents/WWC Pigeon Hole/23-12-12 WOODWARD/23-12-12 papers.pdf)
      /UF (C:/Users/samsu/Dropbox/My documents/WWC Pigeon Hole/23-12-12 WOODWARD/23-12-12 papers.pdf)
      /Type /Filespec
      /NewWindow true
    >>
  >>
  /Rect [ 63 25 570 31 ]
  /BS <<
    /W 0
  >>
  /Subtype /Link
  /NM (fitz-L65)
>>

What I've noticed examining the link created in Evermap is that it is structured differently with a further xref for 'A' and a further xref for 'F' . However, when expanded it looks the same as that following your hack:

A ('xref', '98 0 R')
  D ('array', '[22/XYZ 0 792 0]')
  F ('xref', '99 0 R')
   F ('string', 'buildingofficeaddins.pdf')
   Type ('name', '/Filespec')
   S ('name', '/GoToR')
   NewWindow ('bool', 'true')
  BS ('dict', '<</S/U/Type/Border/W 1>>')
  Border ('array', '[0 0 1]')
  H ('name', '/I')
  M ('string', 'D:20240205110338Z')
  Rect ('array', '[72 756.568 83.077 771.286]')
  Subtype ('name', '/Link')
  Type ('name', '/Annot')

This link works as it should but I can't see any logical difference from the link generated by your hack.

Is it possible to create xrefs programmatically within other xrefs?

**Reply by chapmanwilliam** [reply, marked as answer]
Correction. I think I have fixed it:

            new_doc.xref_set_key(l['xref'],'A/NewWindow','true')

**Reply by chapmanwilliam** [reply]
And for those who want to fix the zoom to fit the page:

            new_doc.xref_set_key(l['xref'], 'A/D', f"[{pg}/Fit]")

**Comment by JorjMcKie** [comment]
But I think we can simply change our link-skeleton to include the `/NewWindow` key with the value `true`. This should meet the vast majority of user intentions, and I do not want to fiddle around with the user interface for such a corner case - at this point in time.

**Reply by chapmanwilliam** [reply]
Sounds good. How do you mean?

**Reply by chapmanwilliam** [reply]
I think I've solved it -above - many thanks for your help.

**Comment by drkousek** [comment]
@chapmanwilliam or @JorjMcKie can you please provide minimal working example?
I am trying to incorporate your things, but they aren't working
